### ISSUE TITLE
Input loses focus if part of update uses IO

### ISSUE THREAD

**Issue by LukaHorvat** [pkamenarsky/concur-replica#45]
This is the problematic widget:

```haskell
test :: Widget HTML a
test = go ""
    where
    go search = do
        s <- div [] [do
            liftIO (pure ())
            -- unsafeBlockingIO (pure ())
            targetValue . target
                <$> input [type_ "text", value search, onInput, Concur.Replica.id "loc"] ]
        go s
```

It's a div wrapping an input field. The widget is recursive and feeds the input back into itself. The problem is that after typing one character, the field loses focus. What happens is that one key press results in two UI updates. The first one deletes everything, the second one rebuilds it which recreates the input field, making it not in focus anymore.

The problem is in the `liftIO (pure ())` line. If it's removed or replaced by the unsafe variant, the UI only gets one update and the field remains focused.

This doesn't happen if we remove the outer div or move the IO action out of the div, but in my case that isn't a workable solution.

Now, since the `unsafeBlockingIO` function exists, I guess this might be intended behavior but in that case I'd like to hear a solution to my issue that doesn't involve using unsafe functions, or at least I'd like a better understanding of just how unsafe the function is.

**Comment by pkamenarsky** [comment]
I *suspect* this is an issue with the underlying `concur-core` - normally, this shouldn't be happening as the VDOM should persist through IO operations until it is explicitly replaced by another widget.

I've been meaning to dive deep and fundamentally rebuild `concur-core`, since conceptually it's actually pretty simple, but I haven't had the time and/or motivation yet. Also, iirc Anupam (the creator of Concur) is working on a new implementation as well, which might solve this and other little problems reported on the issue tracker in one go.

So, tldr, this is definitely a bug, but unfortunately there's no quick fix yet :(

**Comment by seagreen** [comment]
Random idea: it might be cool to have a reference implementation of `concur-core` that only cares about correctness and readability, not speed, so we could use it to check where bugs like this are occurring.

It would also be cool to have around for understanding how `concur-core` works, one could start by reading the reference implementation and only read the real one if they care about the speed tricks.

Just an idea, I'm not  sure if it would be worth the work.
